// @sapphire/plugin-logger/dist/cjs/lib/LoggerStyle.cjs
'use strict';

var Colorette = require('colorette');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var Colorette__namespace = /*#__PURE__*/_interopNamespace(Colorette);

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __publicField = (obj, key, value) => __defNormalProp(obj, key + "" , value);
var _LoggerStyle = class _LoggerStyle {
  constructor(resolvable = {}) {
    __publicField(this, "style");
    if (typeof resolvable === "function") {
      this.style = resolvable;
    } else {
      const styles = [];
      if (resolvable.effects) styles.push(...resolvable.effects.map((text) => Colorette__namespace[text]));
      if (resolvable.text) styles.push(Colorette__namespace[resolvable.text]);
      if (resolvable.background) styles.push(Colorette__namespace[resolvable.background]);
      this.style = styles.length ? styles.length === 1 ? styles[0] : (string) => styles.reduce((out, style) => style(out), string) : Colorette__namespace.reset;
    }
  }
  /**
   * Applies the style to a string.
   * @since 1.0.0
   * @param string The value to apply the style to.
   */
  run(string) {
    return this.style(string);
  }
};
__name(_LoggerStyle, "LoggerStyle");
var LoggerStyle = _LoggerStyle;
var LoggerStyleEffect = /* @__PURE__ */ ((LoggerStyleEffect2) => {
  LoggerStyleEffect2["Reset"] = "reset";
  LoggerStyleEffect2["Bold"] = "bold";
  LoggerStyleEffect2["Dim"] = "dim";
  LoggerStyleEffect2["Italic"] = "italic";
  LoggerStyleEffect2["Underline"] = "underline";
  LoggerStyleEffect2["Inverse"] = "inverse";
  LoggerStyleEffect2["Hidden"] = "hidden";
  LoggerStyleEffect2["Strikethrough"] = "strikethrough";
  return LoggerStyleEffect2;
})(LoggerStyleEffect || {});
var LoggerStyleText = /* @__PURE__ */ ((LoggerStyleText2) => {
  LoggerStyleText2["Black"] = "black";
  LoggerStyleText2["Red"] = "red";
  LoggerStyleText2["Green"] = "green";
  LoggerStyleText2["Yellow"] = "yellow";
  LoggerStyleText2["Blue"] = "blue";
  LoggerStyleText2["Magenta"] = "magenta";
  LoggerStyleText2["Cyan"] = "cyan";
  LoggerStyleText2["White"] = "white";
  LoggerStyleText2["Gray"] = "gray";
  LoggerStyleText2["BlackBright"] = "blackBright";
  LoggerStyleText2["RedBright"] = "redBright";
  LoggerStyleText2["GreenBright"] = "greenBright";
  LoggerStyleText2["YellowBright"] = "yellowBright";
  LoggerStyleText2["BlueBright"] = "blueBright";
  LoggerStyleText2["MagentaBright"] = "magentaBright";
  LoggerStyleText2["CyanBright"] = "cyanBright";
  LoggerStyleText2["WhiteBright"] = "whiteBright";
  return LoggerStyleText2;
})(LoggerStyleText || {});
var LoggerStyleBackground = /* @__PURE__ */ ((LoggerStyleBackground2) => {
  LoggerStyleBackground2["Black"] = "bgBlack";
  LoggerStyleBackground2["Red"] = "bgRed";
  LoggerStyleBackground2["Green"] = "bgGreen";
  LoggerStyleBackground2["Yellow"] = "bgYellow";
  LoggerStyleBackground2["Blue"] = "bgBlue";
  LoggerStyleBackground2["Magenta"] = "bgMagenta";
  LoggerStyleBackground2["Cyan"] = "bgCyan";
  LoggerStyleBackground2["White"] = "bgWhite";
  LoggerStyleBackground2["BlackBright"] = "bgBlackBright";
  LoggerStyleBackground2["RedBright"] = "bgRedBright";
  LoggerStyleBackground2["GreenBright"] = "bgGreenBright";
  LoggerStyleBackground2["YellowBright"] = "bgYellowBright";
  LoggerStyleBackground2["BlueBright"] = "bgBlueBright";
  LoggerStyleBackground2["MagentaBright"] = "bgMagentaBright";
  LoggerStyleBackground2["CyanBright"] = "bgCyanBright";
  LoggerStyleBackground2["WhiteBright"] = "bgWhiteBright";
  return LoggerStyleBackground2;
})(LoggerStyleBackground || {});

exports.LoggerStyle = LoggerStyle;
exports.LoggerStyleBackground = LoggerStyleBackground;
exports.LoggerStyleEffect = LoggerStyleEffect;
exports.LoggerStyleText = LoggerStyleText;
