// @sapphire/framework/dist/esm/preconditions/GuildNewsThreadOnly.mjs
import { __name } from '../chunk-PAWJFY3S.mjs';
import { container } from '@sapphire/pieces';
import { ChannelType } from 'discord.js';
import { Identifiers } from '../lib/errors/Identifiers.mjs';
import { AllFlowsPrecondition } from '../lib/structures/Precondition.mjs';

var _CorePrecondition = class _CorePrecondition extends AllFlowsPrecondition {
  messageRun(message) {
    return message.thread?.type === ChannelType.AnnouncementThread ? this.ok() : this.makeSharedError();
  }
  async chatInputRun(interaction) {
    const channel = await this.fetchChannelFromInteraction(interaction);
    return channel.type === ChannelType.AnnouncementThread ? this.ok() : this.makeSharedError();
  }
  async contextMenuRun(interaction) {
    const channel = await this.fetchChannelFromInteraction(interaction);
    return channel.type === ChannelType.AnnouncementThread ? this.ok() : this.makeSharedError();
  }
  makeSharedError() {
    return this.error({
      identifier: Identifiers.PreconditionGuildNewsThreadOnly,
      message: "You can only run this command in server announcement thread channels."
    });
  }
};
__name(_CorePrecondition, "CorePrecondition");
var CorePrecondition = _CorePrecondition;
void container.stores.loadPiece({
  name: "GuildNewsThreadOnly",
  piece: CorePrecondition,
  store: "preconditions"
});

export { CorePrecondition };
