// @sapphire/framework/dist/esm/arguments/CoreInteger.mjs
import { __name } from '../chunk-PAWJFY3S.mjs';
import { container } from '@sapphire/pieces';
import { Identifiers } from '../lib/errors/Identifiers.mjs';
import { resolveInteger } from '../lib/resolvers/integer.mjs';
import { Argument } from '../lib/structures/Argument.mjs';

var _CoreArgument = class _CoreArgument extends Argument {
  constructor(context) {
    super(context, { name: "integer" });
    this.messages = {
      [Identifiers.ArgumentIntegerTooSmall]: ({ minimum }) => `The given number must be greater than ${minimum}.`,
      [Identifiers.ArgumentIntegerTooLarge]: ({ maximum }) => `The given number must be less than ${maximum}.`,
      [Identifiers.ArgumentIntegerError]: () => "The argument did not resolve to a valid number."
    };
  }
  run(parameter, context) {
    const resolved = resolveInteger(parameter, { minimum: context.minimum, maximum: context.maximum });
    return resolved.mapErrInto(
      (identifier) => this.error({
        parameter,
        identifier,
        message: this.messages[identifier](context),
        context
      })
    );
  }
};
__name(_CoreArgument, "CoreArgument");
var CoreArgument = _CoreArgument;
void container.stores.loadPiece({
  name: "integer",
  piece: CoreArgument,
  store: "arguments"
});

export { CoreArgument };
